// @telegram-apps/init-data-node/dist/entries/node.js
import { createHmac as createHmac$1 } from "node:crypto";
import { h as hashToken$1, i as isValid$1, s as sign$1, a as signData$1, v as validate$1 } from "./shared-aDQj8NQw.js";
import { A, E, d, S, f, g, j, e, c, p, b } from "./shared-aDQj8NQw.js";
function textToBuffer(text) {
  return Buffer.from(typeof text === "string" ? text : new Uint8Array(text));
}
const createHmac = (data, key) => {
  return createHmac$1("sha256", textToBuffer(key)).update(textToBuffer(data)).digest();
};
function hashToken(token) {
  return Buffer.from(hashToken$1(token, createHmac));
}
function isValid(value, token, options) {
  return isValid$1(value, token, validate, options);
}
function sign(data, key, authDate, options) {
  return sign$1(data, key, authDate, signData, options);
}
function signData(data, key, options) {
  return signData$1(data, key, createHmac, options);
}
function validate(value, token, options) {
  return validate$1(value, token, signData, options);
}
export {
  A as AuthDateInvalidError,
  E as ExpiredError,
  d as SignatureInvalidError,
  S as SignatureMissingError,
  hashToken,
  f as isAuthDateInvalidError,
  g as isExpiredError,
  j as isSignatureInvalidError,
  e as isSignatureMissingError,
  isValid,
  c as isValid3rd,
  p as parse,
  sign,
  signData,
  validate,
  b as validate3rd
};
